sourcerer-esjc/src/main/java/org/elder/sourcerer2/esjc/EventStoreEsjcEventRepository.kt

```diff
@@ -0,0 +1,553 @@
+        // We only know the version within the repository if this came from a subscription,
+        // marked by the fact that this is a resolve event (from a projection) rather than a
+        // plain one.
+        val repositoryVersion =
+                if (event.isResolved) event.link.eventNumber
+                else null
+
+        return EventRecord(
+                EventId.ofUuid(event.event.eventId),
+                fromEsStreamId(event.event.eventStreamId),
+                StreamVersion.ofInt(event.event.eventNumber),
+                repositoryVersion?.let { RepositoryVersion.ofInt(it) },
+                event.event.eventType,
+                event.event.created,
+                fromEsMetadata(event.event.metadata),
+                fromEsData(event.event.data))
+    }
+
+    private fun fromEsData(data: ByteArray): T {
+        try {
+            val rawEvent = objectMapper
+                    .readerFor(eventClass)
+                    .readValue<T>(data)
+            return normalizeEvent(rawEvent)
+        } catch (ex: IOException) {
+            throw RetriableEventReadException("Internal error reading events", ex)
+        }
+
+    }
+
+    private fun normalizeEvent(rawEvent: T): T {
+        return if (normalizer != null) {
+            normalizer.normalizeEvent(rawEvent)
+        } else {
+            rawEvent
+        }
+    }
+
+    private fun fromEsMetadata(metadata: ByteArray?): ImmutableMap<String, String> {
+        if (metadata == null || metadata.size == 0) {
+            return ImmutableMap.of()
+        }
+
+        try {
+            return ImmutableMap.copyOf(objectMapper
+                    .readerFor(object : TypeReference<Map<String, String>>() {
+                    })
+                    .readValue<Map<String, String>>(metadata))
+        } catch (ex: IOException) {
+            throw RetriableEventReadException("Internal error reading events", ex)
+        }
+
+    }
+
+    private fun <U> completeReadFuture(future: CompletableFuture<U>): U {
+        try {
+            return future.get(timeoutMillis, TimeUnit.MILLISECONDS)
+        } catch (ex: InterruptedException) {
+            Thread.currentThread().interrupt()
+            throw RetriableEventReadException("Internal error reading event", ex)
+        } catch (ex: ExecutionException) {
+            if (ex.cause is EventStoreException) {
+                if (ex.cause is AccessDeniedException
+                        || ex.cause is CommandNotExpectedException
+                        || ex.cause is InvalidTransactionException
+                        || ex.cause is NoResultException
+                        || ex.cause is NotAuthenticatedException
+                        || ex.cause is PersistentSubscriptionDeletedException
+                        || ex.cause is StreamDeletedException) {
+                    throw PermanentEventReadException(ex.cause)
+                } else if (ex.cause is CannotEstablishConnectionException
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Complex conditions should be simplified and extracted into well-named methods if necessary. SIZE 7 is greater than the threshold 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3AComplexCondition)